src/System.Net.Http/tests/FunctionalTests/HttpClientHandlerTest.ClientCertificates.cs

```diff
@@ -90,15 +90,26 @@ public void ClientCertificateOptions_ValueArg_Roundtrips(ClientCertificateOption
 
             Func<HttpClient, Socket, Uri, X509Certificate2, Task> makeAndValidateRequest = async (client, server, url, cert) =>
             {
-                await TestHelper.WhenAllCompletedOrAnyFailed(
-                    client.GetStringAsync(url),
-                    LoopbackServer.AcceptSocketAsync(server, async (socket, stream, reader, writer) =>
-                    {
-                        SslStream sslStream = Assert.IsType<SslStream>(stream);
-                        Assert.Equal(cert, sslStream.RemoteCertificate);
-                        await LoopbackServer.ReadWriteAcceptedAsync(socket, reader, writer);
-                        return null;
-                    }, options));
+                Task<string> clientTask = client.GetStringAsync(url);
+                Task<List<string>> serverTask = LoopbackServer.AcceptSocketAsync(server, async (socket, stream, reader, writer) =>
+                {
+                    SslStream sslStream = Assert.IsType<SslStream>(stream);
+                    Assert.Equal(cert, sslStream.RemoteCertificate);
+                    socket.NoDelay = true;
+                    var result = await LoopbackServer.ReadWriteAcceptedAsync(socket, reader, writer);
+
+                    // Wait for a bit so client can process response before server goes down.
+                    await Task.Delay(100);
```

Comment:
This doesn't seem like the right fix to me.  Any such delay is still going to be subject to issues in CI / more stressful environments than when just running locally.  But more than that, it seems like it's working some other underlying issue.  It should be possible for the server to write data to the socket and then its half, while still allowing the client to receive the data (in a real server environment, servers don't wait 100ms after sending data to the client).  I suggest the root cause is probably elsewhere, but if we must wait in some fashion, we should instead wait only as long as we need: since the client and server are in the same process, we should be able to easily use a synchronization mechanism (e.g. a TaskCompletionSource) that's set by the client whenever the thing we need is received and that the server can await before closing... again, though, that seems like it's working around an underlying cause we should actually address, unless I'm misunderstanding something fundamental.